# Hands-on: `pandas` & Data Wrangling

By now, you have some experience in using the `pandas` library which will be very helpful in this module. In this notebook, we will explore more of `pandas` but in the context of data wrangling. To be specific, we will be covering the following topics:
- Reading in data
- Descriptive statistics
- Data wrangling
- Filtering
- Aggregation
- Merging

Again we import the necessary libraries first. Always remember to import first.

In [1]:
import pandas as pd
import numpy as np

## Data

The Philippines has an Open Data portal: https://data.gov.ph

In this notebook, we'll be using the [Public Elementary School Enrollment Statistics](https://data.gov.ph/?q=dataset/public-elementary-school-enrollment-statistics) provided by the Department of Education. The page contains two files. Download both files and save them to the same folder as this notebook.

## Reading Data

In the previous modules, we have already demonstrated how to read files using `pandas`. For more details, run the cells below to display the documentations for the commonly used functions for reading files. Try to **read the documentation** to see if what you're trying to do is something that can already done by a library. Or you could simply **google** your concern. Most of the times, someone has already encountered the same problem.

In [2]:
pd.read_csv?

Signature:
pd.read_csv(
    filepath_or_buffer: Union[str, pathlib.Path, IO[~AnyStr]],
    sep=',',
    delimiter=None,
    header='infer',
    names=None,
    index_col=None,
    usecols=None,
    squeeze=False,
    prefix=None,
    mangle_dupe_cols=True,
    dtype=None,
    engine=None,
    converters=None,
    true_values=None,
    false_values=None,
    skipinitialspace=False,
    skiprows=None,
    skipfooter=0,
    nrows=None,
    na_values=None,
    keep_default_na=True,
    na_filter=True,
    verbose=False,
    skip_blank_lines=True,
    parse_dates=False,
    infer_datetime_format=False,
    keep_date_col=False,
    date_parser=None,
    dayfirst=False,
    cache_dates=True,
    iterator=False,
    chunksize=None,
    compression='infer',
    thousands=None,
    decimal: str = '.',
    lineterminator=None,
    quotechar='"',
    quoting=0,
    doublequote=True,
    escapechar=None,
    comment=None,
    encoding=None,
    dialect=None,
    error_bad_lines=True,
    warn_bad_l

In [3]:
pd.read_excel?

Signature:
pd.read_excel(
    io,
    sheet_name=0,
    header=0,
    names=None,
    index_col=None,
    usecols=None,
    squeeze=False,
    dtype=None,
    engine=None,
    converters=None,
    true_values=None,
    false_values=None,
    skiprows=None,
    nrows=None,
    na_values=None,
    keep_default_na=True,
    verbose=False,
    parse_dates=False,
    date_parser=None,
    thousands=None,
    comment=None,
    skipfooter=0,
    convert_float=True,
    mangle_dupe_cols=True,
    **kwds,
)
Docstring:
Read an Excel file into a pandas DataFrame.

Supports `xls`, `xlsx`, `xlsm`, `xlsb`, and `odf` file extensions
read from a local filesystem or URL. Supports an option to read
a single sheet or a list of sheets.

Parameters
----------
io : str, bytes, ExcelFile, xlrd.Book, path object, or file-like object
    Any valid string path is acceptable. The string could be a URL. Valid
    URL schemes include http, ftp, s3, and file. For file URLs, a host is
    expected. A local file coul

In [4]:
# by default, the encoding is utf-8, but since the data has some latin characters
# the encoding argument needs to be updated
# list of encodings can be found here https://docs.python.org/2.4/lib/standard-encodings.html
# read more about encodings here http://kunststube.net/encoding/
deped2012 = pd.read_csv('deped_publicelementaryenrollment2012.csv', encoding='latin1')

# the head function provides a preview of the first 5 rows of the data
deped2012.head()

school_id                    school_name             region      province  \
0     101746                A. Diaz, Sr. ES  I (Ilocos Region)    Pangasinan   
1     102193  A. P. Santos ES (SPED Center)  I (Ilocos Region)  Ilocos Norte   
2     101283               A.P. Guevarra IS  I (Ilocos Region)    Pangasinan   
3     100216                     Ab-Abut ES  I (Ilocos Region)  Ilocos Norte   
4     100043                       Abaca ES  I (Ilocos Region)  Ilocos Norte   

           municipality                  division                district  \
0              Bautista  Pangasinan II, Binalonan                Bautista   
1  Laoag City (Capital)                Laoag City  Laoag City District II   
2             Bayambang    Pangasinan I, Lingayen            Bayambang II   
3                Piddig              Ilocos Norte                  Piddig   
4                Bangui              Ilocos Norte                  Bangui   

  year_level gender  enrollment  
0    grade 1   male          53  
1    grade 1   male          31  
2    grade 1   male          16  
3    grade 1   male          19  
4    grade 1   male          12

In [5]:
# Let's read in the other file too
deped2015 = pd.read_csv('depend_publicelementaryenrollment2015.csv', encoding='latin1')
deped2015.head()

region      province municipality      division  \
0  Region I - Ilocos Region  Ilocos Norte      Bacarra  Ilocos Norte   
1  Region I - Ilocos Region  Ilocos Norte      Bacarra  Ilocos Norte   
2  Region I - Ilocos Region  Ilocos Norte      Bacarra  Ilocos Norte   
3  Region I - Ilocos Region  Ilocos Norte      Bacarra  Ilocos Norte   
4  Region I - Ilocos Region  Ilocos Norte      Bacarra  Ilocos Norte   

   school_id                school_name year_level gender  enrollment  \
0     100001         Apaleng-libtong ES    grade 1   male           9   
1     100002                Bacarra CES    grade 1   male          41   
2     100003                   Buyon ES    grade 1   male           7   
3     100004  Ganagan Elementary School    grade 1   male           8   
4     100005                 Macupit ES    grade 1   male           5   

      latitude    longitude  
0    18.253666    120.60618  
1  18.25096389  120.6089583  
2    18.234599   120.616037  
3  18.25001389  120.5871694  
4  18.29399444  120.6410194

### Let's begin exploring the data...

Some of the most common questions to ask **first** before proceeding with your data is to know the basic details of what's **in** the data. This is an important first step to verify what you see in the preview (`head`) and what's in the entire file.

* How many rows and columns do we have? 
* What is the data type of each column? 
* What is the most common value? Mean? Standard deviation?

#### `shape`

A `pandas` `DataFrame` is essentially a 2D `numpy` array. Using the `shape` attribute of the `DataFrame`, we can easily check the dimensions of the data file we read. It returns a tuple of the dimensions.

In [6]:
deped2012.shape

(463908, 10)

This means that the `deped_publicelementaryenrollment2012.csv` file has 463,908 rows and 10 columns.

#### `dtypes` 
`dtypes` lets you check what data type each column is.

In [7]:
deped2012.dtypes

school_id        int64
school_name     object
region          object
province        object
municipality    object
division        object
district        object
year_level      object
gender          object
enrollment       int64
dtype: object

Notice that everything except `school_id` and `enrollment` is type `object`. In Python, a String is considered an `object`.

#### `describe()`
`describe()` provides the basic descriptive statistics of the`DataFrame`. By default, it only includes the columns with numerical data. Non-numerical columns are omitted but there are arguments that shows the statistics related to non-numerical data.

In [8]:
deped2012.describe()

school_id     enrollment
count  463908.000000  463908.000000
mean   123102.439820      28.582152
std     22010.932343      44.727529
min    100001.000000       0.000000
25%    109747.000000       9.000000
50%    119533.000000      16.000000
75%    129325.000000      31.000000
max    261503.000000    1047.000000

By default we see the **descriptive statistics** of the nnumerical columns.

In [9]:
deped2012.describe(include=np.object)

school_name                  region province municipality division  \
count          463908                  463908   463908       463908   463908   
unique          29699                      17       86         1437      206   
top     San Isidro ES  VIII (Eastern Visayas)    Leyte   Davao City    Leyte   
freq             2328                   43728    15612         3420    14136   

       district year_level  gender  
count    463908     463908  463908  
unique     2415          6       2  
top       Rizal    grade 4    male  
freq       1608      77318  231954

But by specifying the `include` argument, we can see the descriptive statistics of the specific data type we're looking for.

In [10]:
deped2012.describe?

Signature: deped2012.describe(percentiles=None, include=None, exclude=None) -> ~FrameOrSeries
Docstring:
Generate descriptive statistics.

Descriptive statistics include those that summarize the central
tendency, dispersion and shape of a
dataset's distribution, excluding ``NaN`` values.

Analyzes both numeric and object series, as well
as ``DataFrame`` column sets of mixed data types. The output
will vary depending on what is provided. Refer to the notes
below for more detail.

Parameters
----------
percentiles : list-like of numbers, optional
    The percentiles to include in the output. All should
    fall between 0 and 1. The default is
    ``[.25, .5, .75]``, which returns the 25th, 50th, and
    75th percentiles.
include : 'all', list-like of dtypes or None (default), optional
    A white list of data types to include in the result. Ignored
    for ``Series``. Here are the options:

    - 'all' : All columns of the input will be included in the output.
    - A list-like of dtypes

### Data Wrangling

After looking at the basic information about the data, let's see how "clean" the data is

#### Common Data Problems (from slides)
1. Missing values
2. Formatting issues / data types
3. Duplicate records
4. Varying representation / Handle categorical values

#### `isna()` / `isnull()`

To check if there's any missing values, `pandas` provides these two functions to detect them. This actually maps each individual cell to either True or False.

#### `dropna()`

To remove any records with missing values, `dropna()` may be used. It has a number of arguments to help narrow down the criteria for removing the records with missing values.

In [11]:
deped2012.isna?

Signature: deped2012.isna() -> 'DataFrame'
Docstring:
Detect missing values.

Return a boolean same-sized object indicating if the values are NA.
NA values, such as None or :attr:`numpy.NaN`, gets mapped to True
values.
Everything else gets mapped to False values. Characters such as empty
strings ``''`` or :attr:`numpy.inf` are not considered NA values
(unless you set ``pandas.options.mode.use_inf_as_na = True``).

Returns
-------
DataFrame
    Mask of bool values for each element in DataFrame that
    indicates whether an element is not an NA value.

See Also
--------
DataFrame.isnull : Alias of isna.
DataFrame.notna : Boolean inverse of isna.
DataFrame.dropna : Omit axes labels with missing values.
isna : Top-level isna.

Examples
--------
Show which entries in a DataFrame are NA.

>>> df = pd.DataFrame({'age': [5, 6, np.NaN],
...                    'born': [pd.NaT, pd.Timestamp('1939-05-27'),
...                             pd.Timestamp('1940-04-25')],
...                    'name':

In [12]:
deped2012.dropna?

Signature:
deped2012.dropna(
    axis=0,
    how='any',
    thresh=None,
    subset=None,
    inplace=False,
)
Docstring:
Remove missing values.

See the :ref:`User Guide <missing_data>` for more on which values are
considered missing, and how to work with missing data.

Parameters
----------
axis : {0 or 'index', 1 or 'columns'}, default 0
    Determine if rows or columns which contain missing values are
    removed.

    * 0, or 'index' : Drop rows which contain missing values.
    * 1, or 'columns' : Drop columns which contain missing value.

    .. versionchanged:: 1.0.0

       Pass tuple or list to drop on multiple axes.
       Only a single axis is allowed.

how : {'any', 'all'}, default 'any'
    Determine if row or column is removed from DataFrame, when we have
    at least one NA or all NA.

    * 'any' : If any NA values are present, drop that row or column.
    * 'all' : If all values are NA, drop that row or column.

thresh : int, optional
    Require that many non-NA valu

In [13]:
deped2012.isna().sum()

school_id       0
school_name     0
region          0
province        0
municipality    0
division        0
district        0
year_level      0
gender          0
enrollment      0
dtype: int64

In this case, there are no null values which is great, but in most real-world datasets, expect null values.

In [14]:
deped2012_dropped = deped2012.dropna(inplace=False)
deped2012.shape, deped2012_dropped.shape

((463908, 10), (463908, 10))

You'll see above that shape is dimension because nothing happened after applying `dropna` as there are no null values to begin with. But what if there's a null value in this dataset?

In [15]:
# This is just an ILLUSTRATION to show how to handle nan values. Don't change values to NaN unless NEEDED.
deped2012_copy = deped2012.copy()  # We first make a copy of the dataframe
deped2012_copy.iloc[0,0] = np.nan  # We modify the COPY (not the original)
deped2012_copy.head()

school_id                    school_name             region      province  \
0        NaN                A. Diaz, Sr. ES  I (Ilocos Region)    Pangasinan   
1   102193.0  A. P. Santos ES (SPED Center)  I (Ilocos Region)  Ilocos Norte   
2   101283.0               A.P. Guevarra IS  I (Ilocos Region)    Pangasinan   
3   100216.0                     Ab-Abut ES  I (Ilocos Region)  Ilocos Norte   
4   100043.0                       Abaca ES  I (Ilocos Region)  Ilocos Norte   

           municipality                  division                district  \
0              Bautista  Pangasinan II, Binalonan                Bautista   
1  Laoag City (Capital)                Laoag City  Laoag City District II   
2             Bayambang    Pangasinan I, Lingayen            Bayambang II   
3                Piddig              Ilocos Norte                  Piddig   
4                Bangui              Ilocos Norte                  Bangui   

  year_level gender  enrollment  
0    grade 1   male          53  
1    grade 1   male          31  
2    grade 1   male          16  
3    grade 1   male          19  
4    grade 1   male          12

In [16]:
deped2012_copy.isna().sum()

school_id       1
school_name     0
region          0
province        0
municipality    0
division        0
district        0
year_level      0
gender          0
enrollment      0
dtype: int64

There null value is now reflected as shown in the output above

In [17]:
deped2012_dropped = deped2012_copy.dropna(inplace=False)
deped2012_copy.shape, deped2012_dropped.shape

((463908, 10), (463907, 10))

The 'dropped' dataframe now has a lower number of rows compared to the original one.

#### `duplicated()` --> `drop_duplicates()`

The `duplicated()` function returns the duplicated rows in the `DataFrame`. It also has a number of arguments for you to specify the subset of columns. 

`drop_duplicates()` is the function to remove the duplicated rows found by `duplicated()`.

In [18]:
deped2012.duplicated?

Signature:
deped2012.duplicated(
    subset: Union[Hashable, Sequence[Hashable], NoneType] = None,
    keep: Union[str, bool] = 'first',
) -> 'Series'
Docstring:
Return boolean Series denoting duplicate rows.

Considering certain columns is optional.

Parameters
----------
subset : column label or sequence of labels, optional
    Only consider certain columns for identifying duplicates, by
    default use all of the columns.
keep : {'first', 'last', False}, default 'first'
    Determines which duplicates (if any) to mark.

    - ``first`` : Mark duplicates as ``True`` except for the first occurrence.
    - ``last`` : Mark duplicates as ``True`` except for the last occurrence.
    - False : Mark all duplicates as ``True``.

Returns
-------
Series
File:      /anaconda3/envs/dsi/lib/python3.8/site-packages/pandas/core/frame.py
Type:      method


In [19]:
deped2012.drop_duplicates?

Signature:
deped2012.drop_duplicates(
    subset: Union[Hashable, Sequence[Hashable], NoneType] = None,
    keep: Union[str, bool] = 'first',
    inplace: bool = False,
    ignore_index: bool = False,
) -> Union[ForwardRef('DataFrame'), NoneType]
Docstring:
Return DataFrame with duplicate rows removed.

Considering certain columns is optional. Indexes, including time indexes
are ignored.

Parameters
----------
subset : column label or sequence of labels, optional
    Only consider certain columns for identifying duplicates, by
    default use all of the columns.
keep : {'first', 'last', False}, default 'first'
    Determines which duplicates (if any) to keep.
    - ``first`` : Drop duplicates except for the first occurrence.
    - ``last`` : Drop duplicates except for the last occurrence.
    - False : Drop all duplicates.
inplace : bool, default False
    Whether to drop duplicates in place or to return a copy.
ignore_index : bool, default False
    If True, the resulting axis will be

In [20]:
deped2012.duplicated().sum()

0

We can see here that there are no duplicates.

#### Varying representation

For categorical or textual data, unless the options provided are fixed, misspellings and different representations may exist in the same file.

To check the unique values of each column, a `pandas` `Series` has a function `unique()` which returns all the unique values of the column.

In [21]:
deped2012['province'].unique()

array(['Pangasinan', 'Ilocos Norte', 'Ilocos Sur', 'La Union', 'Isabela',
       'Nueva Vizcaya', 'Cagayan', 'Quirino', 'Batanes', 'Nueva Ecija',
       'Bataan', 'Bulacan', 'Aurora', 'Zambales', 'Tarlac', 'Pampanga',
       'Batangas', 'Quezon', 'Laguna', 'Rizal', 'Cavite', 'Palawan',
       'Oriental Mindoro', 'Occidental Mindoro', 'Romblon', 'Marinduque',
       'Camarines Sur', 'Camarines Norte', 'Masbate', 'Sorsogon', 'Albay',
       'Catanduanes', 'Negros Occidental', 'Iloilo', 'Antique', 'Capiz',
       'Aklan', 'Guimaras', 'Bohol', 'Negros Oriental', 'Cebu',
       'Siquijor', 'Eastern Samar', 'Leyte', 'Western Samar',
       'Northern Samar', 'Southern Leyte', 'Biliran', 'Zamboanga del Sur',
       'Zamboanga Sibugay', 'Zamboanga del Norte', 'City of Isabela',
       'Misamis Occidental', 'Lanao del Norte', 'Misamis Oriental',
       'Bukidnon', 'Camiguin', 'Davao del Sur', 'Davao del Norte',
       'Davao Oriental', 'Compostela Valley', 'South Cotabato',
       'Sarangani', '

In [22]:
deped2012['year_level'].unique()

array(['grade 1', 'grade 2', 'grade 3', 'grade 4', 'grade 5', 'grade 6'],
      dtype=object)

In [23]:
deped2012['region'].unique()

array(['I (Ilocos Region)', 'II (Cagayan Valley)', 'III (Central Luzon)',
       'IV-A (CALABARZON)', 'IV-B (MIMAROPA)', 'V (Bicol Region)',
       'VI (Western Visayas)', 'VII (Central Visayas)',
       'VIII (Eastern Visayas)', 'IX (Zamboanga Peninsula)',
       'X (Northern Mindanao)', 'XI (Davao Region)', 'XII (SOCCSKSARGEN)',
       'XIII (Caraga)', 'ARMM (Autonomous Region in Muslim Mindanao)',
       'CAR (Cordillera Administrative Region)',
       'NCR (National Capital Region)'], dtype=object)

In [24]:
deped2015['region'].unique()

array(['Region I - Ilocos Region', 'Region II - Cagayan Valley',
       'Region III - Central Luzon', 'Region IV-A - CALABARZON',
       'Region IV-B - MIMAROPA', 'Region V - Bicol Region',
       'Region VI - Western Visayas', 'Region VII - Central Visayas',
       'Region VIII - Eastern Visayas', 'Region IX - Zamboanga Peninsula',
       'Region X - Northern Mindanao', 'Region XI - Davao Region',
       'Region XII - SOCCSKSARGEN', 'CARAGA - CARAGA',
       'ARMM - Autonomous Region in Muslim Mindanao'], dtype=object)

### Summarizing Data

High data granularity is great for a detailed analysis. However, data is usually summarized or aggregated prior to visualization. `pandas` also provides an easy way to summarize data based on the columns you'd like using the `groupby` function.

We can call any of the following when grouping by columns:
- count()
- sum()
- min()
- max()
- std()

For columns that are categorical in nature, we can simply do `df['column'].value_counts()`. This will give the frequency of each unique value in the column. 

In [25]:
pd.Series.value_counts?

Signature:
pd.Series.value_counts(
    self,
    normalize=False,
    sort=True,
    ascending=False,
    bins=None,
    dropna=True,
)
Docstring:
Return a Series containing counts of unique values.

The resulting object will be in descending order so that the
first element is the most frequently-occurring element.
Excludes NA values by default.

Parameters
----------
normalize : bool, default False
    If True then the object returned will contain the relative
    frequencies of the unique values.
sort : bool, default True
    Sort by frequencies.
ascending : bool, default False
    Sort in ascending order.
bins : int, optional
    Rather than count values, group them into half-open bins,
    a convenience for ``pd.cut``, only works with numeric data.
dropna : bool, default True
    Don't include counts of NaN.

Returns
-------
Series

See Also
--------
Series.count: Number of non-NA elements in a Series.
DataFrame.count: Number of non-NA elements in a DataFrame.

Examples
--------
>>

Number of region instances

In [26]:
deped2015['region'].value_counts()

Region VIII - Eastern Visayas                  41484
Region VI - Western Visayas                    39204
Region V - Bicol Region                        35892
Region VII - Central Visayas                   33768
Region III - Central Luzon                     33336
Region IV-A - CALABARZON                       31452
Region I - Ilocos Region                       27648
Region II - Cagayan Valley                     24924
Region IX - Zamboanga Peninsula                23724
Region X - Northern Mindanao                   23532
Region IV-B - MIMAROPA                         20232
Region XI - Davao Region                       18420
CARAGA - CARAGA                                18408
Region XII - SOCCSKSARGEN                      18228
ARMM - Autonomous Region in Muslim Mindanao     6036
Name: region, dtype: int64

In [27]:
deped2012.groupby?

Signature:
deped2012.groupby(
    by=None,
    axis=0,
    level=None,
    as_index: bool = True,
    sort: bool = True,
    group_keys: bool = True,
    squeeze: bool = False,
    observed: bool = False,
) -> 'groupby_generic.DataFrameGroupBy'
Docstring:
Group DataFrame using a mapper or by a Series of columns.

A groupby operation involves some combination of splitting the
object, applying a function, and combining the results. This can be
used to group large amounts of data and compute operations on these
groups.

Parameters
----------
by : mapping, function, label, or list of labels
    Used to determine the groups for the groupby.
    If ``by`` is a function, it's called on each value of the object's
    index. If a dict or Series is passed, the Series or dict VALUES
    will be used to determine the groups (the Series' values are first
    aligned; see ``.align()`` method). If an ndarray is passed, the
    values are used as-is determine the groups. A label or list of
    labels 

Number of enrollments per grade level

In [28]:
deped2012.groupby("year_level")['enrollment'].sum()

year_level
grade 1    2762527
grade 2    2426057
grade 3    2191076
grade 4    2063463
grade 5    1965628
grade 6    1850738
Name: enrollment, dtype: int64

#### Exercise! 

Let's try to get the following:
1. Total number of enrolled students per region and gender
2. Total number of enrolled students per year level and gender

In [29]:
deped2012.groupby(['region', 'gender'], as_index=False).sum()

region  gender   school_id  \
0   ARMM (Autonomous Region in Muslim Mindanao)  female  1981060044   
1   ARMM (Autonomous Region in Muslim Mindanao)    male  1981060044   
2        CAR (Cordillera Administrative Region)  female  1301046546   
3        CAR (Cordillera Administrative Region)    male  1301046546   
4                             I (Ilocos Region)  female  1481294424   
5                             I (Ilocos Region)    male  1481294424   
6                           II (Cagayan Valley)  female  1396560900   
7                           II (Cagayan Valley)    male  1396560900   
8                           III (Central Luzon)  female  1965169458   
9                           III (Central Luzon)    male  1965169458   
10                            IV-A (CALABARZON)  female  1814452494   
11                            IV-A (CALABARZON)    male  1814452494   
12                              IV-B (MIMAROPA)  female  1271909262   
13                              IV-B (MIMAROPA)    male  1271909262   
14                     IX (Zamboanga Peninsula)  female  1619836632   
15                     IX (Zamboanga Peninsula)    male  1619836632   
16                NCR (National Capital Region)  female   438511446   
17                NCR (National Capital Region)    male   438511446   
18                             V (Bicol Region)  female  2190420960   
19                             V (Bicol Region)    male  2190420960   
20                         VI (Western Visayas)  female  2419755186   
21                         VI (Western Visayas)    male  2419755186   
22                        VII (Central Visayas)  female  2157645486   
23                        VII (Central Visayas)    male  2157645486   
24                       VIII (Eastern Visayas)  female  2754627936   
25                       VIII (Eastern Visayas)    male  2754627936   
26                        X (Northern Mindanao)  female  1648145328   
27                        X (Northern Mindanao)    male  1648145328   
28                            XI (Davao Region)  female  1307008878   
29                            XI (Davao Region)    male  1307008878   
30                           XII (SOCCSKSARGEN)  female  1426850814   
31                           XII (SOCCSKSARGEN)    male  1426850814   
32                                XIII (Caraga)  female  1379807532   
33                                XIII (Caraga)    male  1379807532   

    enrollment  
0       325728  
1       299438  
2       102069  
3       113411  
4       298996  
5       330234  
6       207741  
7       226833  
8       629770  
9       685337  
10      768652  
11      836389  
12      224875  
13      248030  
14      278387  
15      302026  
16      601227  
17      644892  
18      467493  
19      521294  
20      491396  
21      546490  
22      488937  
23      541420  
24      336485  
25      370882  
26      314561  
27      343111  
28      326488  
29      356601  
30      304751  
31      322966  
32      192611  
33      209968

In [30]:
deped2012.groupby(['year_level', 'gender']).sum()

school_id  enrollment
year_level gender                        
grade 1    female  4759017221     1262015
           male    4759017221     1500512
grade 2    female  4759017221     1144869
           male    4759017221     1281188
grade 3    female  4759017221     1052206
           male    4759017221     1138870
grade 4    female  4759017221     1003971
           male    4759017221     1059492
grade 5    female  4759017221      966741
           male    4759017221      998887
grade 6    female  4759017221      930365
           male    4759017221      920373

### Filtering Data



In [31]:
deped2015.query("year_level=='grade 6'")

region       province  \
330240                     Region I - Ilocos Region   Ilocos Norte   
330241                     Region I - Ilocos Region   Ilocos Norte   
330242                     Region I - Ilocos Region   Ilocos Norte   
330243                     Region I - Ilocos Region   Ilocos Norte   
330244                     Region I - Ilocos Region   Ilocos Norte   
...                                             ...            ...   
396283  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur   
396284  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur   
396285  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur   
396286  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur   
396287  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur   

                           municipality            division  school_id  \
330240                          Bacarra        Ilocos Norte     100001   
330241                          Bacarra        Ilocos Norte     100002   
330242                          Bacarra        Ilocos Norte     100003   
330243                          Bacarra        Ilocos Norte     100004   
330244                          Bacarra        Ilocos Norte     100005   
...                                 ...                 ...        ...   
396283  Bacolod-Kalawi (Bacolod Grande)  Lanao del Sur - II     133553   
396284  Bacolod-Kalawi (Bacolod Grande)  Lanao del Sur - II     133554   
396285  Bacolod-Kalawi (Bacolod Grande)  Lanao del Sur - II     133555   
396286  Bacolod-Kalawi (Bacolod Grande)  Lanao del Sur - II     133556   
396287  Bacolod-Kalawi (Bacolod Grande)  Lanao del Sur - II     133557   

                      school_name year_level  gender  enrollment     latitude  \
330240         Apaleng-libtong ES    grade 6    male           3    18.253666   
330241                Bacarra CES    grade 6    male          25  18.25096389   
330242                   Buyon ES    grade 6    male           9    18.234599   
330243  Ganagan Elementary School    grade 6    male           5  18.25001389   
330244                 Macupit ES    grade 6    male           7  18.29399444   
...                           ...        ...     ...         ...          ...   
396283                   Tambo PS    grade 6  female           0            -   
396284                    Tuka PS    grade 6  female           0            -   
396285                  Tulain PS    grade 6  female           0            -   
396286                  Ulodan ES    grade 6  female          52            -   
396287    Awani Elementary School    grade 6  female          40            -   

          longitude  
330240    120.60618  
330241  120.6089583  
330242   120.616037  
330243  120.5871694  
330244  120.6410194  
...             ...  
396283            -  
396284            -  
396285            -  
396286            -  
396287            -  

[66048 rows x 11 columns]

In [32]:
deped2015.query("year_level == 'grade 6' & school_id == 100004")

region      province municipality      division  \
330243  Region I - Ilocos Region  Ilocos Norte      Bacarra  Ilocos Norte   
363267  Region I - Ilocos Region  Ilocos Norte      Bacarra  Ilocos Norte   

        school_id                school_name year_level  gender  enrollment  \
330243     100004  Ganagan Elementary School    grade 6    male           5   
363267     100004  Ganagan Elementary School    grade 6  female           8   

           latitude    longitude  
330243  18.25001389  120.5871694  
363267  18.25001389  120.5871694

In [33]:
deped2015.query("year_level == 'grade 6' | year_level == 'grade 5'")[['region', 'province']]

region       province
264192                     Region I - Ilocos Region   Ilocos Norte
264193                     Region I - Ilocos Region   Ilocos Norte
264194                     Region I - Ilocos Region   Ilocos Norte
264195                     Region I - Ilocos Region   Ilocos Norte
264196                     Region I - Ilocos Region   Ilocos Norte
...                                             ...            ...
396283  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur
396284  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur
396285  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur
396286  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur
396287  ARMM - Autonomous Region in Muslim Mindanao  Lanao Del Sur

[132096 rows x 2 columns]

### Merging Data

Data are sometimes separated into different files or additional data from another source can be associated to another dataset. `pandas` provides means to combine different `DataFrames` together (provided that there are common variables that it can connect them to.

#### `pd.merge`
`merge()` is very similar to database-style joins. `pandas` allows merging of `DataFrame` and **named** `Series` objects together. A join can be done along columns or indexes.

#### `pd.concat`
`concat()` on the other hand combines `pandas` objects along a specific axis.

#### `df.append`
`append()` basically adds the rows of another `DataFrame` or `Series` to the end of the caller. 

In [34]:
pd.merge?

Signature:
pd.merge(
    left,
    right,
    how: str = 'inner',
    on=None,
    left_on=None,
    right_on=None,
    left_index: bool = False,
    right_index: bool = False,
    sort: bool = False,
    suffixes=('_x', '_y'),
    copy: bool = True,
    indicator: bool = False,
    validate=None,
) -> 'DataFrame'
Docstring:
Merge DataFrame or named Series objects with a database-style join.

The join is done on columns or indexes. If joining columns on
columns, the DataFrame indexes *will be ignored*. Otherwise if joining indexes
on indexes or indexes on a column or columns, the index will be passed on.

Parameters
----------
left : DataFrame
right : DataFrame or named Series
    Object to merge with.
how : {'left', 'right', 'outer', 'inner'}, default 'inner'
    Type of merge to be performed.

    * left: use only keys from left frame, similar to a SQL left outer join;
      preserve key order.
    * right: use only keys from right frame, similar to a SQL right outer join;
      pres

In [35]:
pd.concat?

Signature:
pd.concat(
    objs: Union[Iterable[Union[ForwardRef('DataFrame'), ForwardRef('Series')]], Mapping[Union[Hashable, NoneType], Union[ForwardRef('DataFrame'), ForwardRef('Series')]]],
    axis=0,
    join='outer',
    ignore_index: bool = False,
    keys=None,
    levels=None,
    names=None,
    verify_integrity: bool = False,
    sort: bool = False,
    copy: bool = True,
) -> Union[ForwardRef('DataFrame'), ForwardRef('Series')]
Docstring:
Concatenate pandas objects along a particular axis with optional set logic
along the other axes.

Can also add a layer of hierarchical indexing on the concatenation axis,
which may be useful if the labels are the same (or overlapping) on
the passed axis number.

Parameters
----------
objs : a sequence or mapping of Series or DataFrame objects
    If a dict is passed, the sorted keys will be used as the `keys`
    argument, unless it is passed, in which case the values will be
    selected (see below). Any None objects will be dropped silen

In [36]:
deped2012.append?

Signature:
deped2012.append(
    other,
    ignore_index=False,
    verify_integrity=False,
    sort=False,
) -> 'DataFrame'
Docstring:
Append rows of `other` to the end of caller, returning a new object.

Columns in `other` that are not in the caller are added as new columns.

Parameters
----------
other : DataFrame or Series/dict-like object, or list of these
    The data to append.
ignore_index : bool, default False
    If True, do not use the index labels.
verify_integrity : bool, default False
    If True, raise ValueError on creating index with duplicates.
sort : bool, default False
    Sort columns if the columns of `self` and `other` are not aligned.

    .. versionadded:: 0.23.0
    .. versionchanged:: 1.0.0

        Changed to not sort by default.

Returns
-------
DataFrame

See Also
--------
concat : General function to concatenate DataFrame or Series objects.

Notes
-----
If a list of dict/series is passed and the keys are all contained in
the DataFrame's index, the order o

In [37]:
stats2012 = deped2012.groupby('school_id', as_index=False).sum()
stats2015 = deped2015.groupby('school_id', as_index=False).sum()

In [38]:
stats2012.head()

school_id  enrollment
0     100001          72
1     100002         365
2     100003         138
3     100004          89
4     100005          73

In [39]:
stats2015.tail()

school_id  enrollment
33019     133553         142
33020     133554         253
33021     133555         240
33022     133556         333
33023     133557         401

In [40]:
stats2012.append(stats2015)

school_id  enrollment
0         100001          72
1         100002         365
2         100003         138
3         100004          89
4         100005          73
...          ...         ...
33019     133553         142
33020     133554         253
33021     133555         240
33022     133556         333
33023     133557         401

[71683 rows x 2 columns]

#### Exercise

The task is to compare the enrollment statistics of the elementary schools between 2012 and 2015. 

1. Get the total number of enrolled students per school for each year
2. Merge the two `DataFrame`s together to show the summarized statistics for the two years for all schools.

In [41]:
stats2012 = deped2012.groupby('school_id', as_index=False).sum()
stats2015 = deped2015.groupby('school_id', as_index=False).sum()

In [42]:
stats2012.head()

school_id  enrollment
0     100001          72
1     100002         365
2     100003         138
3     100004          89
4     100005          73

In [43]:
stats2012.shape

(38659, 2)

In [44]:
stats2015.head()

school_id  enrollment
0     100001          72
1     100002         407
2     100003         152
3     100004         107
4     100005          67

In [45]:
stats2015.shape

(33024, 2)

The following is the wrong way of merging this.

In [46]:
merged = pd.merge(stats2012, stats2015)
merged.head()

school_id  enrollment
0     100001          72
1     100021          39
2     100040         116
3     100044          14
4     100155         201

In [47]:
merged.shape

(526, 2)

#### Observations

1. Are the number of rows for both `DataFrames` the same or different? What's the implication if they're different?
2. Note the same column names for the two `DataFrames`. Based on the documentation for `merge()`, there's a parameter for suffixes for overlapping column names. If we want to avoid the "messy" suffixes, we can choose to rename columns prior to merging.

One way is to assign an array to the columns object representing the column names for ALL columns.

```ipython
stats2012.columns = ['school_id', '2012']
stats2015.columns = ['school_id', '2015']
```

But this is not good if you have too many columns... `pandas` has a function `rename()` in which we can pass a "mappable" dictionary for the columns. The `inplace` parameter helps in renaming it and assigns the changed `DataFrame` back to the same variable.

```ipython
stats2012.rename(columns={'enrollment': '2012'}, inplace=True)
stats2015.rename(columns={'enrollment': '2015'}, inplace=True)
```

In [48]:
# try the code above
stats2012.columns = ['school_id', '2012']
stats2015.columns = ['school_id', '2015']

In [49]:
stats2012.head()

school_id  2012
0     100001    72
1     100002   365
2     100003   138
3     100004    89
4     100005    73

In [50]:
stats2015.head()

school_id  2015
0     100001    72
1     100002   407
2     100003   152
3     100004   107
4     100005    67

In [51]:
## Merge the two dataframes using different "how" parameters
# how : {'left', 'right', 'outer', 'inner'}, default 'inner'

In [52]:
inner_res = pd.merge(stats2012, stats2015)
inner_res.head()

school_id  2012  2015
0     100001    72    72
1     100002   365   407
2     100003   138   152
3     100004    89   107
4     100005    73    67

In [53]:
inner_res.isna().sum()

school_id    0
2012         0
2015         0
dtype: int64

In [54]:
inner_res.shape

(32998, 3)

Play around with the how parameter and observe the following: 
- shape of the dataframe 
- presence or absence of null values 
- number of schools dropped with respect to the original dataframe

In [55]:
outer_res = pd.merge(stats2012, stats2015, how="outer")
outer_res.isna().sum()

school_id       0
2012           26
2015         5661
dtype: int64

In [56]:
left_res = pd.merge(stats2012, stats2015, how="left")
left_res.isna().sum()

school_id       0
2012            0
2015         5661
dtype: int64

For the following items, we will only be using the 2015 dataset.

1. Which region has the most number of schools? Does this region also have the most number of enrollees?

In [57]:
deped2015.groupby(['region']).sum().sort_values(by='enrollment', ascending=False)

school_id  enrollment
region                                                             
Region IV-A - CALABARZON                     3413404572     1568707
Region III - Central Luzon                   3527190888     1259020
Region VI - Western Visayas                  4558487448     1025647
Region VII - Central Visayas                 4029909312     1009273
Region V - Bicol Region                      4059563592      918017
Region XI - Davao Region                     2376901548      677463
Region VIII - Eastern Visayas                5081872632      656566
Region X - Northern Mindanao                 2994705924      652688
Region XII - SOCCSKSARGEN                    2380915416      613998
Region I - Ilocos Region                     2796762168      612984
Region IX - Zamboanga Peninsula              2971818840      547674
Region IV-B - MIMAROPA                       2240173944      442229
Region II - Cagayan Valley                   2576090532      435115
CARAGA - CARAGA                              2433546996      393040
ARMM - Autonomous Region in Muslim Mindanao   804444624      131314

2. Which region has the least number of schools? Does this region also have the least number of enrollees?

In [58]:
deped2015.groupby(['region']).sum().sort_values(by='enrollment', ascending=True)

school_id  enrollment
region                                                             
ARMM - Autonomous Region in Muslim Mindanao   804444624      131314
CARAGA - CARAGA                              2433546996      393040
Region II - Cagayan Valley                   2576090532      435115
Region IV-B - MIMAROPA                       2240173944      442229
Region IX - Zamboanga Peninsula              2971818840      547674
Region I - Ilocos Region                     2796762168      612984
Region XII - SOCCSKSARGEN                    2380915416      613998
Region X - Northern Mindanao                 2994705924      652688
Region VIII - Eastern Visayas                5081872632      656566
Region XI - Davao Region                     2376901548      677463
Region V - Bicol Region                      4059563592      918017
Region VII - Central Visayas                 4029909312     1009273
Region VI - Western Visayas                  4558487448     1025647
Region III - Central Luzon                   3527190888     1259020
Region IV-A - CALABARZON                     3413404572     1568707

3. Which school has the least number of enrollees? 

In [59]:
deped2015.groupby(['school_name']).sum().sort_values(by='enrollment', ascending=True)

school_id  enrollment
school_name                                            
Lamitan City Sped Center            1595880           0
Basakan PS                          1598160           0
Rufino Elementary School            1476480           0
Subah Languyan PS                   1597548           0
Maluno Integrated School Annex-A    1237344           0
...                                     ...         ...
San Roque ES                      125530512       27673
Sta. Cruz ES                      102643704       28250
San Vicente ES                    142484232       33621
San Jose ES                       174943476       43098
San Isidro ES                     222516348       54815

[25543 rows x 2 columns]